core/storage/src/storages/beacon.rs

```diff
@@ -22,6 +39,175 @@ impl GenericStorage<SignedBeaconBlockHeader, SignedBeaconBlock> for BeaconChainS
+            processed_blocks: Default::default(),
+            thresholds: Default::default(),
+            accepted_authorities: Default::default(),
+        }
     }
-}
+
+    /// whether there is authority info in storage
+    pub fn is_authority_empty(&self) -> bool {
+        let proposals: HashMap<_, _> = self.generic_storage.storage.iter(Some(COL_PROPOSAL)).collect();
+        proposals.is_empty()
+    }
+
+    pub fn prune_authority_storage(
+        &mut self,
+        slot_filter: &Fn(Slot) -> bool,
+        epoch_filter: &Fn(Epoch) -> bool,
+    ) {
+        prune_index(
+            self.generic_storage.storage.as_ref(),
+            COL_PROPOSAL,
+            &mut self.proposals,
+            slot_filter,
+        ).expect("Failed to prune storage");
+        prune_index(
+            self.generic_storage.storage.as_ref(),
+            COL_PARTICIPATION,
+            &mut self.participation,
+            slot_filter,
+        ).expect("Failed to prune storage");
+        prune_index(
+            self.generic_storage.storage.as_ref(),
+            COL_PROCESSED_BLOCKS,
+            &mut self.processed_blocks,
+            epoch_filter,
+        ).expect("Failed to prune storage");
+        prune_index(
+            self.generic_storage.storage.as_ref(),
+            COL_THRESHOLD,
+            &mut self.thresholds,
+            epoch_filter,
+        ).expect("Failed to prune storage");
+        prune_index(
+            self.generic_storage.storage.as_ref(),
+            COL_ACCEPTED_AUTHORITY,
+            &mut self.accepted_authorities,
+            slot_filter,
+        ).expect("Failed to prune storage");
+    }
+
+    pub fn get_proposal(&mut self, slot: Slot) -> Option<&Vec<AuthorityStake>> {
+        read_with_cache(
+            self.generic_storage.storage.as_ref(),
+            COL_PROPOSAL,
+            &mut self.proposals,
+            &self.generic_storage.enc_index(slot)
+        ).expect("Failed to read from storage")
+    }
+
+    pub fn set_proposal(&mut self, slot: Slot, proposal: Vec<AuthorityStake>) {
+        write_with_cache(
+            self.generic_storage.storage.as_ref(),
+            COL_PROPOSAL,
+            &mut self.proposals,
+            &self.generic_storage.enc_index(slot),
+            proposal
+        ).expect("Failed to write to storage")
+    }
+
+    pub fn get_participation(&mut self, slot: Slot) -> Option<&AuthorityMask> {
+        read_with_cache(
+            self.generic_storage.storage.as_ref(),
+            COL_PARTICIPATION,
+            &mut self.participation,
+            &self.generic_storage.enc_index(slot)
+        ).expect("Failed to read from storage")
+    }
+
+    pub fn set_participation(&mut self, slot: Slot, mask: AuthorityMask) {
+        write_with_cache(
+            self.generic_storage.storage.as_ref(),
+            COL_PARTICIPATION,
+            &mut self.participation,
+            &self.generic_storage.enc_index(slot),
```

Comment:
I just realized that we might not be able to implement `Borrow<[u8]>` for `Slot` and `Epoch`, because `Slot` and `Epoch` are type aliases of the types outside our crates.

Reply:
True. We can wrap them in new types if we really want to, but then there need to be more boilerplate code when converting from/to slices.

Reply:
Yeah, let's not do this for now.